src/herder/TransactionQueue.cpp

```diff
@@ -0,0 +1,228 @@
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "herder/TransactionQueue.h"
+#include "crypto/SecretKey.h"
+#include "ledger/LedgerTxn.h"
+#include "main/Application.h"
+#include "transactions/TransactionUtils.h"
+#include "util/HashOfHash.h"
+#include "util/XDROperators.h"
+
+#include <algorithm>
+
+namespace stellar
+{
+
+static uint64_t
+countTxs(TransactionQueue::AccountTxMap const& acc)
+{
+    uint64_t sz = 0;
+    for (auto const& a : acc)
+    {
+        sz += a.second->mTransactions.size();
+    }
+    return sz;
+}
+
+static std::shared_ptr<TransactionQueue::TxMap>
+findOrAdd(TransactionQueue::AccountTxMap& acc, AccountID const& aid)
+{
+    std::shared_ptr<TransactionQueue::TxMap> txmap = nullptr;
+    auto i = acc.find(aid);
+    if (i == acc.end())
+    {
+        txmap = std::make_shared<TransactionQueue::TxMap>();
+        acc.insert(std::make_pair(aid, txmap));
+    }
+    else
+    {
+        txmap = i->second;
+    }
+    return txmap;
+}
+
+void
+TransactionQueue::TxMap::addTx(TransactionFramePtr tx)
+{
+    auto const& h = tx->getFullHash();
+    if (mTransactions.find(h) != mTransactions.end())
+    {
+        return;
+    }
+    mTransactions.insert(std::make_pair(h, tx));
+    mCurrentState.mMaxSeq = std::max(tx->getSeqNum(), mCurrentState.mMaxSeq);
+    mCurrentState.mTotalFees += tx->getFeeBid();
+}
+
+void
+TransactionQueue::TxMap::recalculate()
+{
+    mCurrentState.mMaxSeq = 0;
+    mCurrentState.mTotalFees = 0;
+    for (auto const& pair : mTransactions)
+    {
+        mCurrentState.mMaxSeq =
+            std::max(pair.second->getSeqNum(), mCurrentState.mMaxSeq);
+        mCurrentState.mTotalFees += pair.second->getFeeBid();
+    }
+}
+
+TransactionQueue::TransactionQueue(Application& app, int pendingDepth)
+    : mApp(app), mPendingTransactions(pendingDepth)
+{
+}
+
+TransactionSubmitStatus
+TransactionQueue::add(TransactionFramePtr tx)
+{
+    if (contains(tx))
+    {
+        return TransactionSubmitStatus::DUPLICATE;
+    }
+
+    auto state = getState(tx->getSourceID());
+    LedgerTxn ltx(mApp.getLedgerTxnRoot());
+    if (!tx->checkValid(ltx, state.mMaxSeq))
+    {
+        return TransactionSubmitStatus::ERROR;
+    }
+
+    auto sourceAccount = stellar::loadAccount(ltx, tx->getSourceID());
+    if (getAvailableBalance(ltx.loadHeader(), sourceAccount) <
```

Comment:
This may overflow, replace with 
```
if (getAvailableBalance(ltx.loadHeader(), sourceAccount)-tx->getFeeBid() <
        state.mTotalFees)
```

Reply:
(I think with this change, the rest of the code is fine as this is the only place where we add transactions)